modules/fbx/fbx_parser/FBXError.h

```diff
@@ -0,0 +1,68 @@
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2021 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2021 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+//
+// Created by Gordon MacPherson on 20/04/2021.
+//
+
+#ifndef FBX_ERROR_H
+#define FBX_ERROR_H
+
+#include "core/string/print_string.h"
+
+#define FBX_ERROR_DETECTED FBXError::IsCorrupt()
+#define IF_FBX_IS_CORRUPT_RETURN \
+	if (FBXError::IsCorrupt()) { \
+		return;                  \
+	}
+#define FBX_CORRUPT FBXError::SetCorrupt()
+#define FBX_CORRUPT_ERROR_PTR \
+	FBX_CORRUPT;              \
+	return nullptr;
+
+struct FBXError {
+	static bool IsCorrupt() {
+		return FBXError::corrupt;
+	}
+
+	static void ClearCorrupt() {
+		FBXError::corrupt = false;
+	}
+
+	static void SetCorrupt() {
+		print_error("FBX Document was found to be corrupt, we have decided to stop all operations");
+		FBXError::corrupt = true;
+	}
+
+	// NOTE: thread local should be required, but may crash since godot is not in the same static lib
+	thread_local static bool corrupt;
```

Comment:
One thing I dislike is that some compilers crash with thread_local when it is used in general. I think though it should be OK here, we can try it, if it crashes we need to work out what compilers have issues with thread_local and then look at fixing it or disallowing usage of thread local. We experienced crashes from doctest in the past when this was used, and the workaround was to disable the thread_local storag specifier.

The intention is to make a single copy of the data per thread, this should be correct for FBX since it can only use 1 thread per imported file, then the other files being imported on different threads should import correctly.

Reply:
Poke @hpvb @RandomShaper.